### PR TITLE
[BPK-1025] Refactor card component

### PR DESCRIPTION
**Pull request by matthewdavidson**
I'm working on the new ticket component, which is essentially a copy of the card component. When I looked at the card component I noticed things I wanted to change. I'll comment on the changes inline.

### CODE REVIEW

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 1]
```diff
@@ -57,22 +57,17 @@ const {
 const styles = StyleSheet.create({
   card: {
     backgroundColor: colorWhite,
+    borderRadius: borderRadiusSm,
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: `borderRadius` is not common and is only required on the outer card container.

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 2]
```diff
@@ -57,22 +57,17 @@ const {
...
     elevation: elevationXs,
     shadowColor: shadowSmColor,
     shadowOffset: { width: shadowSmOffsetWidth, height: shadowSmOffsetHeight / PixelRatio.get() },
     shadowOpacity: shadowSmOpacity,
     shadowRadius: shadowSmRadius / PixelRatio.get(),
   },
-  common: {
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: Removed `common` styles as they were redundant.

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 3]
```diff
@@ -57,22 +57,17 @@ const {
...
-    borderRadius: borderRadiusSm,
-  },
-  overflowHidden: {
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: Tested this on iOS & android - this style seemed to be redundant. 

Popped some extra long words to try and force an overflow ... nothing happened. Happy to be proven wrong..?

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 4]
```diff
@@ -83,40 +78,34 @@ const styles = StyleSheet.create({
...
-  const cardBaseStyle = [styles.card, styles.common];
-
-  if (padded) {
-    cardBaseStyle.push(styles.padded);
-  }
-
-  if (focused) {
-    cardBaseStyle.push(styles.focused);
-  }
+  const style = [styles.card];
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: renamed `cardBaseStyle` to `style` - we are in the Card component we don;t need to prefix local variable with `card`.

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 5]
```diff
@@ -83,40 +78,34 @@ const styles = StyleSheet.create({
...
+  if (padded) { style.push(styles.cardPadded); }
+  if (focused) { style.push(styles.cardFocused); }
+  if (userStyle) { style.push(userStyle); }
 
   return (
     <TouchableHighlight
       accessibilityComponentType="button"
-      onPress={onPress}
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: `onPress` is handled by `{...rest}`

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 6]
```diff
@@ -83,40 +78,34 @@ const styles = StyleSheet.create({
...
-      style={[cardBaseStyle, style]}
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: Use *one* mechanism to merge styles. Either if statements above the JSX or rely on react natives ability to flatten arrays - not both.

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 7]
```diff
@@ -83,40 +78,34 @@ const styles = StyleSheet.create({
...
       underlayColor={colorGray50}
+      style={style}
       {...rest}
     >
-      <View
-        style={[styles.overflowHidden, styles.common]}
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: Removed redundant styles.

File: native/packages/react-native-bpk-component-card/stories.js [hunk 8]
```diff
@@ -1,73 +1,59 @@
...
 
+const content = (
+  <BpkText>
+    Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
+    commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus
+    et magnis dis parturient montes, nascetur ridiculus mus.
+  </BpkText>
+);
 
-const tokens = Platform.select({
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/stories.js]: No need to import tokens - just rely on flex box and space around.

File: native/packages/react-native-bpk-component-card/stories.js [hunk 9]
```diff
@@ -1,73 +1,59 @@
...
-  ios: () => require('bpk-tokens/tokens/ios/base.react.native.common.js'), // eslint-disable-line global-require
-  android: () => require('bpk-tokens/tokens/android/base.react.native.common.js'), // eslint-disable-line global-require
-})();
+const onPress = action('Card press');
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/stories.js]: Use storybooks `action` to log events to browser console.

File: native/packages/react-native-bpk-component-card/stories.js [hunk 10]
```diff
@@ -1,73 +1,59 @@
...
-    marginBottom: tokens.spacingBase,
+  allCardsContainer: {
+    flex: 1,
+    flexDirection: 'column',
+    justifyContent: 'space-around',
   },
 });
 
-const cardContent = (
-  <View>
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/stories.js]: Removed extraneous `<View>` wrapper.

File: native/packages/react-native-bpk-component-card/stories.js [hunk 11]
```diff
@@ -1,73 +1,59 @@
...
-  container: {
-    marginBottom: tokens.spacingBase,
+  allCardsContainer: {
+    flex: 1,
+    flexDirection: 'column',
+    justifyContent: 'space-around',
   },
 });
 
-const cardContent = (
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/stories.js]: `cardContent` can just be `content`

File: native/packages/react-native-bpk-component-card/stories.js [hunk 12]
```diff
@@ -1,73 +1,59 @@
...
   .add('docs:without-padding', () => (
-    <BpkCard onPress={press} padded={false}>{cardContent}</BpkCard>
+    <BpkCard onPress={onPress} padded={false}>{content}</BpkCard>
   ))
   .add('docs:focused', () => (
-    <BpkCard onPress={press} focused>{cardContent}</BpkCard>
+    <BpkCard onPress={onPress} focused>{content}</BpkCard>
   ))
   .add('All Cards', () => (
-    <View>
```
Comment (matthewdavidson) [inline native/packages/react-native-bpk-component-card/stories.js]: Removed extraneous `<View>` wrapper.

File: native/packages/react-native-bpk-component-card/src/BpkCard.js [hunk 13]
```diff
@@ -83,40 +78,34 @@ const styles = StyleSheet.create({
...
+      <View>{children}</View>
     </TouchableHighlight>
   );
 };
 
 BpkCard.propTypes = {
   children: PropTypes.node.isRequired,
-  focused: PropTypes.bool,
   onPress: PropTypes.func.isRequired,
+  focused: PropTypes.bool,
```
Comment (matteo-hertel) [inline native/packages/react-native-bpk-component-card/src/BpkCard.js]: they were in alphabetical order, not anymore 😢 but required are on top, mixed feelings here

Comment (matthewdavidson) [reply native/packages/react-native-bpk-component-card/src/BpkCard.js]: if you look at the rest of the codebase, `required` are always on top. We should stick to conventions.

Comment (matteo-hertel) [reply native/packages/react-native-bpk-component-card/src/BpkCard.js]: yes I know, that's why mixed feelings, it's in line with the codebase but not alphabetical anymore 😄

**Comment by matthewdavidson** [comment]
@matteo-hertel agreed - wanna walk through it together?

File: native/packages/react-native-bpk-component-card/src/BpkCard.js (hunk 3 above)
Comment (matthewdavidson) [reply native/packages/react-native-bpk-component-card/src/BpkCard.js]: Ah actually this style mattered for the inner `<View>` and `padded={false}` as it's corners would bleed outwith the outer container. Very interesting ... Setting `color: 'transparent'` also works here... worth keeping in mind in terms of a gotcha.

cc @k0nserv, @shaundon, @matteo-hertel, @georgegillams

Comment (k0nserv) [reply native/packages/react-native-bpk-component-card/src/BpkCard.js]: We should document in code when we do things that "seem" wrong but have particular purpose

Comment (shaundon) [reply native/packages/react-native-bpk-component-card/src/BpkCard.js]: @k0nserv agreed - we should always make it clear when we're doing something unusual as to what's happening :)

**Pull request merged**